napari/layers/utils/style_encoding.py

```diff
@@ -0,0 +1,228 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterable, Optional, Sequence, Tuple, Union
+
+import numpy as np
+from pydantic import ValidationError, parse_obj_as, validator
+
+from ...utils import Colormap
+from ...utils.colormaps import ValidColormapArg, ensure_colormap
+from ...utils.colormaps.categorical_colormap import CategoricalColormap
+from ...utils.colormaps.standardize_color import transform_color
+from ...utils.events import EventedModel
+
+
+class StyleEncoding(EventedModel, ABC):
+    array: np.ndarray = []
+
+    @abstractmethod
+    def _apply(
+        self, properties: Dict[str, np.ndarray], indices: Sequence[int]
+    ) -> np.ndarray:
+        pass
+
+    def update_all(self, properties: Dict[str, np.ndarray], n_rows: int):
+        indices = range(0, n_rows)
+        self.array = self._apply(properties, indices)
+
+    def update_tail(self, properties: Dict[str, np.ndarray], n_rows: int):
+        n_values = self.array.shape[0]
+        indices = range(n_values, n_rows)
```

Comment:
am i being dumb here, but if `n_rows=5` and `n_values=20` say, as there are more values than rows, doesn't this look like `list(range(20, 5)) = []` i.e. nothing? it's late right now so i might be extra dumb in this comment, apologies!!

Reply:
I think this is resolved by the comments above, but clearly I need to do at least change some names to make this less confusing.